usagov/usagov-2021
# Usagov 1214 fix credentials script generation

## Jira Task

https://cm-jira.usa.gov/browse/USAGOV-1214
https://cm-jira.usa.gov/browse/USAGOV-1299

## Description

This is a fix of the implementation of address validation in the CEO tool.

Things that were added to the code:
- Fix the creation of USPS credentials so that it works locally and in other development environments (dev, stage, prod). I added commands in the CircleCI configuration to add the credentials to the env file. I also moved the commands that add the credentials to a JS script to /scripts/bootstrap.sh. 
- Fix CEO error messages in Spanish (I Merge Isabel's change into this branch)
- Fix CEO behavior for new error message from the USPS API. 

## Type of Changes

  - Add screenshot showing what it should look like

## Testing Instructions



**Add the id for the Address Suggestion alert box:**
1. Go to http://localhost/es/funcionarios-electos-resultados
2. Click the "Edit" button 
3. Add `id="address-heading"` to the "Su dirección:" heading.
<img width="864" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/84413c4a-a82c-48db-b956-841c12c51696">

4.  Click "Save".

####
**Check if the USPS API validator isn't working, we should still give them the results for their unvalidated address:**
1. Run `docker compose up`.
2. Check if the script "web/themes/custom/usagov/scripts/usps-credentials.js" was generated and it only contains this: `const error = 'No credentials found in the env.'`.
3. Open localhost and flush all caches.
4. Go to http://localhost/es/funcionarios-electos
5. Fill out the form using any address you want. Here are the values that I used:
** Dirección**: 20043 El Ray Lane
**Ciudad**: Sonora
**Estado**: California
**Código postal**: 95370
6. Click the button to submit the form.
7. Check that you're getting the results without any errors and without the address suggestion alert. It should look something like this:
<img width="659" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/17bedd9c-7900-41cc-a128-5e901726d7d5">
  
####
**Check if the credentials are added:**
1. Find your env.local file and add this environment variables:
```
USPS_USERID=
USPS_PASSWORD=
```
Note: If you don't have the credentials, please send me a message. Please, do not share the credentials or publish them on GitHub.
2. Run `docker compose up`.
3. Using Explorer in VSCode, check if the script "web/themes/custom/usagov/scripts/usps-credentials.js" was generated and contains the credentials like this:
```
const USPS_USERID = '...';
const USPS_PASSWORD = '...';
```

####  
**USPS API returns "Address Not Found.":**

1. Run `docker compose up`. 
2. Open localhost and flush all caches .
3. Go to http://localhost/es/funcionarios-electos.
5. Fill out the form using these values:
** Dirección**: El Ray Lane
**Ciudad**: Sonora
**Estado**: California
**Código postal**: 95370
7. Click the button to submit the form.
8. Check that you're getting the "Dirección no encontrada. Por favor, escriba una dirección válida." error. It should look something like this:
<img width="491" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/493fd30b-50a3-4459-882c-e8797166f1a9">

####
**USPS API returns "Invalid City.":**

1. Go to http://localhost/es/funcionarios-electos.
2. Fill out the form using these values:
** Dirección**: 20043 El Ray Lane
**Ciudad**: SSS
**Estado**: California
**Código postal**: 12345
6. Click the button to submit the form.
7. Check that you're getting the "Ciudad no encontrada. Por favor, escriba una ciudad válida." error. It should look something like this:
<img width="499" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/bf6875d1-f8a2-40d9-a494-33efc369ac37">

####
**USPS API returns "Invalid Zip Code.":**

1. Go to http://localhost/es/funcionarios-electos.
2. Fill out the form using these values:
** Dirección**: 20043 El Ray Lane
**Ciudad**: SSS
**Estado**: California
**Código postal**: 11111
6. Click the button to submit the form.
7. Check that you're getting the "Por favor, escriba un código postal válido de 5 dígitos." error. It should look something like this:
<img width="492" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/3878f2f1-5c0c-41c0-b7df-60dab1aa2c24">

####
**USPS API returns address suggestion:**

1. Go to http://localhost/es/funcionarios-electos
2. Fill out the form using these values:
** Dirección**: 1 river place
**Ciudad**: New York
**Estado**: New York
**Código postal**: 10128
3. Click the button to submit the form.
4. Check that you're getting the Address Suggestion Alert Box. It should look something like this:
<img width="461" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/e315a3d1-5a47-428b-977d-ddd7249c462a">

5. Click the link "Use this address for your search" and check if the pages look something like this:
<img width="495" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/1e38d4f6-bcbf-4b32-8d15-a9ef7c3ed4be">

####
**Test New USPS API error Message:**

1. Go to http://localhost/es/funcionarios-electos
2. Fill out the form using these values:
** Dirección**: Chicago Skyway
**Ciudad**: Chicago
**Estado**: Illinois
**Código postal**: 60602
3. Click the button to submit the form.
4. Check if the page look something like this:
<img width="512" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/96c31c93-8635-485f-af41-fec9e41b8419">

### Requires New Config

### Requires New Content

### Validation Steps

## Security Review




## Reviewer Reminders
- Reviewed code changes
- Reviewed functionality
- Security review complete or not required

## Post PR Approval Instructions
Follow these steps as soon as you merge the new changes.

1. Go to the [USAGov Circle CI project](https://app.circleci.com/pipelines/github/usagov/usagov-2021).
14. Find the commit of this pull request.
15. Build and deploy the changes.
16. Update the Jira ticket by changing the ticket status to `Review in Test` and add a comment. State whether the change is already visible on [cms-dev.usa.gov](http://cms-dev.usa.gov/) and [beta-dev.usa.gov](http://beta-dev.usa.gov/), or if the deployment is still in process.
